server/backend/api/app/crud/services/permission_audit_service.py

```diff
@@ -0,0 +1,172 @@
+        self.repo.save(audit_record)
+
+    def store_user_permissions_revoked_audit_history(
+        self, requester: RequesterSchema, delete_record: FamUserRoleXref
+    ):
+        revoked_permission_target_user = delete_record.user
+        revoked_permission_role = delete_record.role
+        change_type = PrivilegeChangeTypeEnum.REVOKE
+        audit_record = PermissionAduitHistoryCreateSchema(
+            application_id=revoked_permission_role.application.application_id,
+            create_user=requester.user_name,
+            change_performer_user_id=requester.user_id,
+            change_target_user_id=revoked_permission_target_user.user_id,
+            change_performer_user_details=PermissionAuditService.to_change_performer_user_details(requester),
+            privilege_change_type_code=change_type,
+            privilege_details=PermissionAuditService.to_enduser_privliege_revoked_details(delete_record),
+            create_date=datetime.datetime.now(datetime.UTC),
+            change_date=datetime.datetime.now(datetime.UTC)
+        )
+
+        LOGGER.debug(f"Adding audit record for ({change_type}): {audit_record}")
+        self.repo.save(audit_record)
+
+    @staticmethod
+    def to_change_performer_user_details(
+        requester: RequesterSchema
+    ) -> PrivilegeChangePerformerSchema:
+        return PrivilegeChangePerformerSchema(
+            username=requester.user_name,
+            first_name=requester.first_name,
+            last_name=requester.last_name,
+            email=requester.email
+        )
+
+    @staticmethod
+    def to_enduser_privliege_granted_details(enduser_privliege_list: List[FamUserRoleAssignmentCreateRes]):
+        if (len(enduser_privliege_list) == 0):
+            return
+
+        # Note, current FAM supports creating ONLY 1 role with multiple forest_client(s) at ONLY 1 scope type ("CLIENT").
+        # TODO !! When team begins new scoped roles model, it needs refactoring.  # noqa NOSONAR
+        def __map_to_privilege_role_scope(item: FamUserRoleAssignmentCreateRes):
+            assigned_role = item.detail.role
+            is_forest_client_scoped_role = assigned_role.forest_client is not None
+            return PrivilegeDetailsScopeSchema(
+                scope_type=PrivilegeDetailsScopeTypeEnum.CLIENT,
+                client_id=assigned_role.forest_client.forest_client_number if is_forest_client_scoped_role else None,
+                client_name=assigned_role.forest_client.client_name if is_forest_client_scoped_role else None
+            )
+
+        scopes = list(map(__map_to_privilege_role_scope, enduser_privliege_list))
+        return PrivilegeDetailsSchema(
+            permission_type=PrivilegeDetailsPermissionTypeEnum.END_USER,
+            roles=[
+                PrivilegeDetailsRoleSchema(
+                    role=enduser_privliege_list[0].detail.role.display_name,
+                    scopes=scopes
+                )
+            ]
+        )
```

Comment:
This is only convenient method for transformation between schema objects to build the end user `enduser_privliege_granted_details`